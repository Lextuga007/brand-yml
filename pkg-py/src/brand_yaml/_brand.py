from __future__ import annotations

from typing import Any, List, Literal, Union

from pydantic import BaseModel, ConfigDict, HttpUrl, RootModel


class Brand(BaseModel):
    model_config = ConfigDict(extra="ignore")

    meta: BrandMeta = None
    logo: str | BrandLogo = None
    color: BrandColor = None
    typography: BrandTypography = None
    defaults: dict[str, Any] = None

class BrandStringLightDark(BaseModel):
    model_config = ConfigDict(extra="ignore", str_strip_whitespace=True)

    light: str | dict = None
    dark: str | dict = None


class BrandLogo(BaseModel):
    model_config = ConfigDict(extra="forbid")

    with_: dict[str, str | BrandStringLightDark] = None
    small: str | BrandStringLightDark = None
    medium: str | BrandStringLightDark = None
    large: str | BrandStringLightDark = None


type BrandColorValue = str


class BrandColor(BaseModel):
    model_config = ConfigDict(extra="forbid")

    with_: dict[str, BrandColorValue] = None
    foreground: BrandColorValue = None
    background: BrandColorValue = None
    primary: BrandColorValue = None
    secondary: BrandColorValue = None
    tertiary: BrandColorValue = None
    success: BrandColorValue = None
    info: BrandColorValue = None
    warning: BrandColorValue = None
    danger: BrandColorValue = None
    light: BrandColorValue = None
    dark: BrandColorValue = None
    emphasis: BrandColorValue = None
    link: BrandColorValue = None


type BrandNamedColorType = Literal[
    "foreground",
    "background",
    "primary",
    "secondary",
    "tertiary",
    "success",
    "info",
    "warning",
    "danger",
    "light",
    "dark",
    "emphasis",
    "link",
]
BrandNamedColorValues: BrandNamedColorType = (
    "foreground",
    "background",
    "primary",
    "secondary",
    "tertiary",
    "success",
    "info",
    "warning",
    "danger",
    "light",
    "dark",
    "emphasis",
    "link",
)


class BrandTypography(BaseModel):
    with_: BrandFont | list[BrandFont] = None
    base: BrandTypographyOptions = None
    headings: BrandTypographyOptionsNoSize = None
    monospace: BrandTypographyOptions = None
    emphasis: BrandTypographyEmphasis = None
    link: BrandTypographyLink = None


class BrandTypographyOptions(BaseModel):
    family: str = None
    size: str = None
    line_height: str = None
    weight: BrandTypographyFontWeightType | list[BrandTypographyFontWeightType] = None
    style: BrandTypographyFontStyleType | list[BrandTypographyFontStyleType] = None
    color: str | BrandNamedColorType = None
    background_color: str | BrandNamedColorType = None


class BrandTypographyOptionsNoSize(BaseModel):
    family: str = None
    line_height: str = None
    weight: BrandTypographyFontWeightType | list[BrandTypographyFontWeightType] = None
    style: BrandTypographyFontStyleType | list[BrandTypographyFontStyleType] = None
    color: str | BrandNamedColorType = None
    background_color: str | BrandNamedColorType = None


class BrandTypographyEmphasis(BaseModel):
    weight: BrandTypographyFontWeightType | list[BrandTypographyFontWeightType] = None
    color: str | BrandNamedColorType = None
    background_color: str | BrandNamedColorType = None


class BrandTypographyLink(BaseModel):
    weight: BrandTypographyFontWeightType | list[BrandTypographyFontWeightType] = None
    decoration: str = None
    color: str | BrandNamedColorType = None
    background_color: str | BrandNamedColorType = None


type BrandTypographyFontWeightType = Literal[
    100, 200, 300, 400, 500, 600, 700, 800, 900
]
BrandTypographyFontWeightValues: BrandTypographyFontWeightType = (
    100,
    200,
    300,
    400,
    500,
    600,
    700,
    800,
    900,
)


type BrandTypographyFontStyleType = Literal["normal", "italic"]
BrandTypographyFontStyleValues: BrandTypographyFontStyleType = ("normal", "italic")


class BrandTypographyFontGoogle(BaseModel):
    family: str
    weight: BrandTypographyFontWeightType | list[BrandTypographyFontWeightType] = 400
    style: BrandTypographyFontStyleType | list[BrandTypographyFontStyleType] = "normal"
    display: Literal["auto", "block", "swap", "fallback", "optional"] = "auto"


class BrandFontFile(BaseModel):
    model_config = ConfigDict(extra="forbid")
    family: str
    files: str | list[str] = None


class BrandFontFamily(RootModel):
    root: str


class BrandFont(RootModel):
    root: List[Union[BrandTypographyFontGoogle, BrandFontFile, BrandFontFamily]]
