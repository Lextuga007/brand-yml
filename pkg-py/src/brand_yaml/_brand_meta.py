from __future__ import annotations

from pydantic import BaseModel, ConfigDict, HttpUrl, Field


class BrandMeta(BaseModel):
    """
    Brand metadata is stored in `meta`, providing place to describe the company
    or project, the brand guidelines, additional links, and more.
    """

    model_config = ConfigDict(extra="allow", str_strip_whitespace=True)

    name: str | BrandMetaName = Field(
        None, examples=["Very Big Corporation of America"]
    )
    link: HttpUrl | BrandLink = Field(
        None,
        examples=[
            "https://very-big-corp.com",
            '{"home": "https://very-big-corp.com"}',
        ],
    )


class BrandMetaName(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    full: str = Field(None, examples=["Very Big Corporation of America"])
    short: str = Field(None, examples=["VBC"])


class BrandLink(BaseModel):
    model_config = ConfigDict(extra="allow", str_strip_whitespace=True)

    home: HttpUrl = Field(
        None,
        examples=["https://very-big-corp.com"],
    )
    mastodon: HttpUrl = Field(
        None,
        examples=["https://mastodon.social/@VeryBigCorpOfficial"],
    )
    github: HttpUrl = Field(
        None,
        examples=["https://github.com/Very-Big-Corp"],
    )
    linkedin: HttpUrl = Field(
        None,
        examples=["https://linkedin.com/company/very-big-corp"],
    )
    twitter: HttpUrl = Field(
        None,
        examples=["https://twitter.com/VeryBigCorp"],
    )
    facebook: HttpUrl = Field(
        None,
        examples=["https://facebook.com/Very-Big-Corp"],
    )
