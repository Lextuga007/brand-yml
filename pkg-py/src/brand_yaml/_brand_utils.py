from __future__ import annotations

import logging
from copy import deepcopy

from pydantic import BaseModel, ConfigDict, Field
from typing import Any, TypeVar, Generic, Optional

T = TypeVar("T")


class BrandLightDark(BaseModel, Generic[T]):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    light: T = None
    dark: T = None


class BrandLightDarkString(BrandLightDark[str]):
    pass


class BrandWith(BaseModel, Generic[T]):
    model_config = ConfigDict(
        extra="ignore",
        str_strip_whitespace=True,
        populate_by_name=True,
        revalidate_instances="always",
    )

    with_: Optional[dict[str, T]] = Field(default=None, alias="with")

    def model_post_init(self, __context: Any) -> None:
        if self.with_ is None:
            return

        logging.debug("resolving with_ values")
        self._replace_with_recursively()

    def __setattr__(self, name: str, value: Any) -> None:
        super().__setattr__(name, value)
        self._replace_with_recursively()

    def _replace_with_recursively(self, items: dict | BaseModel | None = None, level=0):
        if level > 50:
            logging.error("BrandWith recursion limit reached")
            return

        if items is None:
            items = self

        if not isinstance(items, (dict, BaseModel)):
            return

        if isinstance(items, BaseModel):
            items_keys = items.model_fields.keys()
        elif hasattr(items, "keys"):
            items_keys = items.keys()

        for key in items_keys:
            logging.debug(f"checking key {key}")

            if isinstance(items, BaseModel):
                value = getattr(items, key)
            elif isinstance(items, dict):
                value = items[key]

            if isinstance(value, str) and value in self.with_:
                logging.debug(f"replacing key {key}")
                if isinstance(items, BaseModel):
                    setattr(items, key, deepcopy(self.with_[value]))
                elif isinstance(items, dict):
                    items[key] = deepcopy(self.with_[value])
            elif isinstance(value, (dict, BaseModel)):
                # TODO: we may want to avoid recursing into child BrandWith instances
                logging.debug(f"recursing into {key}")
                self._replace_with_recursively(value, level + 1)
            else:
                logging.debug(
                    f"skipping {key}, not replaceable (or not a dict or pydantic model)"
                )
