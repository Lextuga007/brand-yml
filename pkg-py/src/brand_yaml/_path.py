from __future__ import annotations

from pathlib import Path

from pydantic import HttpUrl, RootModel


class FileLocation(RootModel):
    root: HttpUrl | Path
    _root_dir: Path

    def __init__(self, path: str | Path | HttpUrl):
        super().__init__(path)
        self._root_dir = Path(".").absolute()

    def __call__(self) -> Path | HttpUrl:
        if isinstance(self.root, Path):
            if self.root.is_absolute():
                return self.root
            return self._root_dir / self.root
        return self.root

    def _update_root_dir(self, root_dir: Path) -> bool:
        self._root_dir = root_dir
        return False

    def _validate_path_exists(self) -> bool:
        path = self()
        if not path or not isinstance(path, Path):
            return False

        if not path.exists():
            raise FileNotFoundError(f"File not found: {path}")

        return False
