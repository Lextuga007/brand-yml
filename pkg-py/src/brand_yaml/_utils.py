from __future__ import annotations

from pathlib import Path


def find_project_file(filename: str, dir_: Path) -> Path:
    dir_og = dir_
    i = 0
    max_parents = 20

    while dir_ != dir_.parent and i < max_parents:
        if (dir_ / filename).exists():
            return dir_ / filename
        dir_ = dir_.parent
        i += 1

    raise FileNotFoundError(
        f"Could not find {filename} in {dir_og} or its parents."
    )
