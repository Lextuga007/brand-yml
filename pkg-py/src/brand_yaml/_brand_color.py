from __future__ import annotations

from typing import Optional

from pydantic import ConfigDict, Field

from ._brand_utils import BrandWith


class BrandColor(BrandWith[str]):
    """
    Brand Colors

    The brand's custom color palette and theme.

    Attributes
    ----------

    foreground
        The foreground color, used for text.
    
    background
        The background color, used for the page or main background.
    
    primary
        The primary accent color, i.e. the main theme color. Typically used for
        hyperlinks, active states, primary action buttons, etc.
    
    secondary
        The secondary accent color. Typically used for lighter text or disabled
        states.
    
    tertiary
        The tertiary accent color. Typically an even lighter color, used for
        hover states, accents, and wells.
    
    success
        The color used for positive or successful actions and information.
    
    info
        The color used for neutral or informational actions and information.
    
    warning
        The color used for warning or cautionary actions and information.
    
    danger
        The color used for errors, dangerous actions, or negative information.
    
    light
        A bright color, used as a high-contrast foreground color on dark
        elements or low-contrast background color on light elements.
    
    dark
        A dark color, used as a high-contrast foreground color on light elements
        or high-contrast background color on light elements.
    
    emphasis
        A color used to emphasize or highlight text or elements.
    
    link
        The color used for hyperlinks. If not defined, the `primary` color is
        used.

    """
    model_config = ConfigDict(extra="forbid")

    foreground: Optional[str] = Field(
        default=None,
        description="The foreground color, used for text.",
    )
    background: Optional[str] = Field(
        default=None,
        description="The background color, used for the page or main background.",
    )
    primary: Optional[str] = Field(
        default=None,
        description="The primary accent color, i.e. the main theme color. Typically used for hyperlinks, active states, primary action buttons, etc.",
    )
    secondary: Optional[str] = Field(
        default=None,
        description="The secondary accent color. Typically used for lighter text or disabled states.",
    )
    tertiary: Optional[str] = Field(
        default=None,
        description="The tertiary accent color. Typically an even lighter color, used for hover states, accents, and wells.",
    )

    success: Optional[str] = Field(
        default=None,
        description="The color used for positive or successful actions and information.",
    )
    info: Optional[str] = Field(
        default=None,
        description="The color used for neutral or informational actions and information.",
    )
    warning: Optional[str] = Field(
        default=None,
        description="The color used for warning or cautionary actions and information.",
    )
    danger: Optional[str] = Field(
        default=None,
        description="The color used for errors, dangerous actions, or negative information.",
    )

    light: Optional[str] = Field(
        default=None,
        description="A bright color, used as a high-contrast foreground color on dark elements or low-contrast background color on light elements.",
    )
    dark: Optional[str] = Field(
        default=None,
        description="A dark color, used as a high-contrast foreground color on light elements or high-contrast background color on light elements.",
    )
    emphasis: Optional[str] = Field(
        default=None,
        description="A color used to emphasize or highlight text or elements.",
    )
    link: Optional[str] = Field(
        default=None,
        description="The color used for hyperlinks. If not defined, the `primary` color is used.",
    )
