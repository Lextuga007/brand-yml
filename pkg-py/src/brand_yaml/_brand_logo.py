from __future__ import annotations

from typing import Union

from pydantic import ConfigDict
from ._brand_utils import BrandStringLightDark, BrandWith

class BrandLogo(BrandWith[Union[str, BrandStringLightDark]]):
    model_config = ConfigDict(extra="forbid")

    small: str | BrandStringLightDark = None
    medium: str | BrandStringLightDark = None
    large: str | BrandStringLightDark = None

