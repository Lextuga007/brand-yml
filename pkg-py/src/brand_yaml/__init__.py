from __future__ import annotations

from pathlib import Path
from typing import Any, Literal, overload

from pydantic import BaseModel, ConfigDict, Field, model_validator
from ruamel.yaml import YAML

from ._path import FileLocation
from ._utils import find_project_brand_yaml, recurse_dicts_and_models
from .color import BrandColor
from .logo import BrandLogo
from .meta import BrandMeta
from .typography import BrandTypography

yaml = YAML()


class Brand(BaseModel):
    model_config = ConfigDict(
        extra="ignore",
        revalidate_instances="always",
        validate_assignment=True,
    )

    meta: BrandMeta | None = Field(None)
    logo: str | BrandLogo | None = Field(None)
    color: BrandColor | None = Field(None)
    typography: BrandTypography | None = Field(None)
    defaults: dict[str, Any] | None = Field(None)
    path: Path | None = Field(None, exclude=True, repr=False)

    @classmethod
    def from_yaml(cls, path: str | Path):
        """
        Read a brand YAML file

        Reads a brand YAML file or finds and reads a `_brand.yml` file and returns
        a validated :class:`Brand` object.

        Parameters
        ----------
        path
            The path to the brand YAML file or a directory where `_brand.yml` is
            expected to be found. Typically, you can pass `__file__` from the
            calling script to find `_brand.yml` in the current directory or any of
            its parent directories.

        Returns
        -------
        :
            A validated :class:`Brand` object with all fields populated according to
            the brand YAML file.

        Raises
        ------
        :
            Raises a `FileNotFoundError` if no brand configuration file is found
            within the given path. Raises `ValueError` or other validation errors
            from [pydantic](https://docs.pydantic.dev/latest/) if the brand YAML
            file is invalid.

        Examples
        --------

        ```python
        from brand_yaml import Brand

        brand = Brand.from_yaml(__file__)
        brand = Brand.from_yaml("path/to/_brand.yml")
        ```
        """
        return cls.model_validate(read_brand_yaml(path, as_data=True))

    @model_validator(mode="after")
    def resolve_typography_colors(self):
        if self.typography is None or self.color is None:
            return self

        color_defs = self.color._color_defs(resolved=True)

        for top_field in self.typography.model_fields.keys():
            typography_node = getattr(self.typography, top_field)

            if not isinstance(typography_node, BaseModel):
                continue

            for typography_node_field in typography_node.model_fields.keys():
                if typography_node_field not in ("color", "background_color"):
                    continue

                value = getattr(typography_node, typography_node_field)
                if value is None or not isinstance(value, str):
                    continue

                if value not in color_defs:
                    continue

                setattr(
                    typography_node,
                    typography_node_field,
                    color_defs[value],
                )

        return self

    @model_validator(mode="after")
    def resolve_paths(self):
        path = self.path
        if path is not None:
            recurse_dicts_and_models(
                self,
                pred=lambda value: isinstance(value, FileLocation),
                modify=lambda value: value.set_root_dir(
                    path.parent,
                    validate_path=True,
                ),
            )
        return self


@overload
def read_brand_yaml(
    path: str | Path, as_data: Literal[False] = False
) -> Brand: ...


@overload
def read_brand_yaml(path: str | Path, as_data: Literal[True]) -> dict: ...


def read_brand_yaml(path: str | Path, as_data: bool = False) -> Brand | dict:
    """
    Read a brand YAML file

    Reads a brand YAML file or finds and reads a `_brand.yml` file and returns
    a validated :class:`Brand` object.

    Parameters
    ----------
    path
        The path to the brand YAML file or a directory where `_brand.yml` is
        expected to be found. Typically, you can pass `__file__` from the
        calling script to find `_brand.yml` in the current directory or any of
        its parent directories.

    as_data
        When `True`, returns the raw brand data as a dictionary parsed from the
        YAML file. When `False`, returns a validated :class:`Brand` object.

    Returns
    -------
    :
        A validated :class:`Brand` object with all fields populated according to
        the brand YAML file (`as_data=False`, default) or the raw brand data
        as a dictionary (`as_data=True`).

    Raises
    ------
    :
        Raises a `FileNotFoundError` if no brand configuration file is found
        within the given path. Raises `ValueError` or other validation errors
        from [pydantic](https://docs.pydantic.dev/latest/) if the brand YAML
        file is invalid.

    Examples
    --------

    ```python
    from brand_yaml import read_brand_yaml

    brand = read_brand_yaml(__file__)
    brand = read_brand_yaml("path/to/_brand.yml")
    ```
    """

    path = Path(path).absolute()

    if path.is_dir():
        path = find_project_brand_yaml(path)
    elif path.suffix == ".py":
        # allows users to simply pass `__file__`
        path = find_project_brand_yaml(path.parent)

    with open(path, "r") as f:
        brand_data = yaml.load(f)

    if not isinstance(brand_data, dict):
        raise ValueError(
            f"Invalid brand YAML file {str(path)!r}. Must be a dictionary."
        )

    brand_data["path"] = path

    if as_data:
        return brand_data

    return Brand.model_validate(brand_data)


__all__ = [
    "Brand",
    "read_brand_yaml",
]
