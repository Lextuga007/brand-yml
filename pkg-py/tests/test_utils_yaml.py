from __future__ import annotations

from brand_yaml import Brand


def test_brand_model_dump_yaml(snapshot):
    brand = Brand.from_yaml_str("""
    meta:
      name: Brand YAML
    color:
      palette:
        orange: "#ff9a02"
      primary: orange
    typography:
      headings: Raleway
    """)

    assert snapshot == brand.model_dump_yaml()
