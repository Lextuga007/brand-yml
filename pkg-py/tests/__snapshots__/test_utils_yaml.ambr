# serializer version: 1
# name: test_brand_model_dump_yaml
  '''
  meta:
    name:
      full: Brand YAML
  color:
    palette:
      orange: '#ff9a02'
    primary: '#ff9a02'
  typography:
    fonts:
    - family: Raleway
      weight:
      - 100
      - 200
      - 300
      - 400
      - 500
      - 600
      - 700
      - 800
      - 900
      url: https://fonts.googleapis.com/
    headings:
      family: Raleway
  
  '''
# ---
